x/genutil/types/genesis.go

```diff
@@ -89,35 +88,50 @@ func (ag *AppGenesis) SaveAs(file string) error {
 
 // AppGenesisFromReader reads the AppGenesis from the reader.
 func AppGenesisFromReader(reader io.Reader) (*AppGenesis, error) {
-	jsonBlob, err := io.ReadAll(reader)
-	if err != nil {
-		return nil, err
-	}
-
-	var appGenesis AppGenesis
-	if err := json.Unmarshal(jsonBlob, &appGenesis); err != nil {
-		// fallback to CometBFT genesis
-		var ctmGenesis cmttypes.GenesisDoc
-		if err2 := cmtjson.Unmarshal(jsonBlob, &ctmGenesis); err2 != nil {
-			return nil, fmt.Errorf("error unmarshalling AppGenesis: %w\n failed fallback to CometBFT GenDoc: %w", err, err2)
+	var ag AppGenesis
+	var err error
+	// check if io.ReadSeeker is implemented
+	if rs, ok := reader.(io.ReadSeeker); ok {
+		err = json.NewDecoder(rs).Decode(&ag)
+		if err == nil {
+			return &ag, nil
 		}
 
-		appGenesis = AppGenesis{
-			AppName: version.AppName,
-			// AppVersion is not filled as we do not know it from a CometBFT genesis
-			GenesisTime:   ctmGenesis.GenesisTime,
-			ChainID:       ctmGenesis.ChainID,
-			InitialHeight: ctmGenesis.InitialHeight,
-			AppHash:       ctmGenesis.AppHash,
-			AppState:      ctmGenesis.AppState,
-			Consensus: &ConsensusGenesis{
-				Validators: ctmGenesis.Validators,
-				Params:     ctmGenesis.ConsensusParams,
-			},
+		err = fmt.Errorf("error unmarshalling AppGenesis: %w", err)
+		if _, serr := rs.Seek(0, io.SeekStart); serr != nil {
+			err = errors.Join(err, fmt.Errorf("error seeking back to the front: %w", serr))
+			return nil, err
 		}
 	}
 
-	return &appGenesis, nil
+	// TODO: once cmtjson implements incremental parsing, we can avoid storing the entire file in memory
+	jsonBlob, ioerr := io.ReadAll(reader)
```

Comment:
`io.ReadAll` buffers the content of the file fronted by `reader` fully in memory right?  If so I don't see how this closes the issue that was linked.

Reply:
I've implemented the suggested changes in the linked issue as is and, depending on the JSON decoder implementation, the only time we're not putting the entire file in memory is when a `io.ReadSeeker` is implemented and the file is decoded successfully (we exit at line 97). Having the code structured this way also let's us easily optimize this further when `cmtjson` supports parsing from a `Reader`.

Reply:
I see.  `os.File` implements `io.ReaderSeeker` so won't this early out (on line 97) without decoding? Has it been tested?

Reply:
In other words, we have 2 code paths fulfilling the same behavior, which is confusing.  I don't understand what `cmtjson` (what is the full package path you are referencing) has to do with things?

Reply:
In my understanding, calling `json.Decode` with an `io.ReadSeeker` should not result in the full file being loaded into memory, instead the file is buffered in to fill the output data structure (`AppGenesis` in our case). So before the memory cost would've been `sizeof(AppGenesis) + sizeof(input_file)` but now we have `sizeof(AppGenesis) + sizeof(input_file_buffer)` which can be a significant when dealing with large files. I can write a novel program that uses `io.ReadAll` and another one that uses `io.ReadSeeker` and compare the memory usage of the two. But I'm not sure if this is the best way, maybe you have other ideas?

`cmtjson` refers to `github.com/cometbft/cometbft/libs/json` and the 2 code paths is not something new I added in this PR, it was there before: if `encoding/json.Decode` failed it'd fallback to `cmtjson.Decode`.

Reply:
ping @kocubinski. Any comment on this?

Reply:
I tried the running:
```go
func main() {
	filename := "large_file.json"
	generateLargeJSONFile(filename, 10_000_000) // Generate a 10M record JSON file

	fmt.Println("Testing json.Unmarshal method:")
	testUnmarshal(filename)

	fmt.Println("\nTesting json.Decoder method:")
	testDecoder(filename)
}

func testUnmarshal(filename string) {
	runtime.GC() // Run garbage collection before starting
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	startAlloc := m.TotalAlloc

	startTime := time.Now()

	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	data, err := io.ReadAll(file)
	if err != nil {
		panic(err)
	}

	var result []LargeStruct
	err = json.Unmarshal(data, &result)
	if err != nil {
		panic(err)
	}

	duration := time.Since(startTime)

	runtime.ReadMemStats(&m)
	memoryUsed := m.TotalAlloc - startAlloc

	fmt.Printf("Time taken: %v\n", duration)
	fmt.Printf("Memory used: %d bytes\n", memoryUsed)
	fmt.Printf("Number of records: %d\n", len(result))
}

func testDecoder(filename string) {
	runtime.GC() // Run garbage collection before starting
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	startAlloc := m.TotalAlloc

	startTime := time.Now()

	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	decoder := json.NewDecoder(file)

	var result []LargeStruct
	err = decoder.Decode(&result)
	if err != nil {
		panic(err)
	}

	duration := time.Since(startTime)

	runtime.ReadMemStats(&m)
	memoryUsed := m.TotalAlloc - startAlloc

	fmt.Printf("Time taken: %v\n", duration)
	fmt.Printf("Memory used: %d bytes\n", memoryUsed)
	fmt.Printf("Number of records: %d\n", len(result))
}
```
and got
```
Testing json.Unmarshal method:
Time taken: 5.009264667s
Memory used: 6010656208 bytes
Number of records: 10000000

Testing json.Decoder method:
Time taken: 4.948757458s
Memory used: 4487049520 bytes
Number of records: 10000000
```